homeassistant/components/tuya/fan.py

```diff
@@ -0,0 +1,259 @@
+        async_add_entities(entities)
+
+    entry.async_on_unload(
+        async_dispatcher_connect(
+            hass, TUYA_DISCOVERY_NEW.format(DEVICE_DOMAIN), async_discover_device
+        )
+    )
+
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    device_ids = []
+    for (device_id, device) in device_manager.device_map.items():
+        if device.category in TUYA_SUPPORT_TYPE:
+            device_ids.append(device_id)
+    async_discover_device(device_ids)
+
+
+def _setup_entities(
+    hass: HomeAssistant, entry: ConfigEntry, device_ids: list[str]
+) -> list[TuyaHaFan]:
+    """Set up Tuya Fan."""
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    entities = []
+    for device_id in device_ids:
+        device = device_manager.device_map[device_id]
+        if device is None:
+            continue
+        entities.append(TuyaHaFan(device, device_manager))
+        hass.data[DOMAIN][entry.entry_id][TUYA_HA_DEVICES].add(device_id)
+    return entities
+
+
+class TuyaHaFan(TuyaHaEntity, FanEntity):
+    """Tuya Fan Device."""
+
+    def __init__(self, device: TuyaDevice, device_manager: TuyaDeviceManager) -> None:
+        """Init Tuya Fan Device."""
+        super().__init__(device, device_manager)
+
+        self.ha_preset_modes = []
+        if DPCODE_MODE in self.tuya_device.function:
+            self.ha_preset_modes = json.loads(
+                self.tuya_device.function[DPCODE_MODE].values
+            ).get("range", [])
+
+        # Air purifier fan can be controlled either via the ranged values or via the enum.
+        # We will always prefer the enumeration if available
+        #   Enum is used for e.g. MEES SmartHIMOX-H06
+        #   Range is used for e.g. Concept CA3000
+        self.air_purifier_speed_range_len = 0
+        self.air_purifier_speed_range_enum = []
+        if self.tuya_device.category == "kj" and (
+            DPCODE_AP_FAN_SPEED_ENUM in self.tuya_device.function
+            or DPCODE_AP_FAN_SPEED in self.tuya_device.function
+        ):
+            if DPCODE_AP_FAN_SPEED_ENUM in self.tuya_device.function:
+                self.dp_code_speed_enum = DPCODE_AP_FAN_SPEED_ENUM
+            else:
+                self.dp_code_speed_enum = DPCODE_AP_FAN_SPEED
+
+            data = json.loads(
+                self.tuya_device.function[self.dp_code_speed_enum].values
+            ).get("range")
+            if data:
+                self.air_purifier_speed_range_len = len(data)
+                self.air_purifier_speed_range_enum = data
+
+    def set_preset_mode(self, preset_mode: str) -> None:
+        """Set the preset mode of the fan."""
+        self._send_command([{"code": DPCODE_MODE, "value": preset_mode}])
+
+    def set_direction(self, direction: str) -> None:
+        """Set the direction of the fan."""
+        self._send_command([{"code": DPCODE_FAN_DIRECTION, "value": direction}])
```

Comment:
Do we need to translate the Home Assistant direction, `DIRECTION_FORWARD` or `DIRECTION_REVERSE`, to a tuya direction before sending the command?

Reply:
The value is the same, no translate is required

Reply:
Ok, let's run with this for now. For the future it's good to add a mapping between the Home Assistant value and the Tuya value to guard the integration for changes in the device API.